src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/text-submission-viewer/text-submission-viewer.component.ts

```diff
@@ -201,47 +201,86 @@ export class TextSubmissionViewerComponent implements OnChanges {
         return this.matches.has(file);
     }
 
-    insertToken(text: string, token: string, position: number) {
-        // prevent negative values because slice does not handle them as we would wish
-        if (position < 0) {
-            position = 0;
+    insertMatchTokens(fileContent: string): string {
+        const matches = this.getMatchesForCurrentFile().sort((m1, m2) => {
+            if (!m1.from || !m1.to || !m2.from || !m2.to) {
+                return 0;
+            }
+            const lines = m1.from.line - m2.from.line;
+            if (lines === 0) {
+                return m1.from.column - m2.from.column;
+            }
+            return lines;
+        });
+        if (!matches.length || !matches[0].from || !matches[0].to) {
+            return escape(fileContent);
         }
-        return [text.slice(0, position), token, text.slice(position)].join('');
-    }
 
-    insertMatchTokens(fileContent: string) {
         const rows = fileContent.split('\n');
-        const matches = this.getMatchesForCurrentFile();
-        const offsets = new Array(rows.length).fill(0);
+        let result = '';
+
+        for (let i = 0; i < matches[0].from.line - 1; i++) {
+            if (rows[i]) {
+                result += escape(rows[i]) + '\n';
+            }
+        }
+        result += escape(rows[matches[0].from.line - 1].slice(0, matches[0].from.column - 1));
 
-        matches.forEach((match) => {
+        for (let i = 0; i < matches.length; i++) {
+            const match = matches[i];
             if (!match.from) {
                 captureException(new Error('"from" is not defined in insertMatchTokens'));
-                return;
+                continue;
             }
             if (!match.to) {
                 captureException(new Error('"to" is not defined in insertMatchTokens'));
-                return;
+                continue;
             }
 
             const idxLineFrom = match.from.line - 1;
             const idxLineTo = match.to.line - 1;
 
-            const idxColumnFrom = match.from.column - 1 + offsets[idxLineFrom];
+            const idxColumnFrom = match.from.column - 1;
+            const idxColumnTo = match.to.column + match.to.length - 1;
 
-            if (rows[idxLineFrom]) {
-                rows[idxLineFrom] = this.insertToken(rows[idxLineFrom], this.tokenStart, idxColumnFrom);
-                offsets[idxLineFrom] += this.tokenStart.length;
+            if (!rows[idxLineFrom]) {
```

Comment:
In don't get this check somehow. So if the first line of the match is empty you skip the entire match. Why? Technically, a match could start in an empty line with ``column === 0``, can it?

Reply:
This check was already in the code, I just inverted it here. But I agree, I don't see why this is needed.